### PR TITLE
Refactor of chapter 6 and edits to chapter 5

### PR DESCRIPTION
**Pull request by jrgold**
Changes to chapter 6:
Removed most introductory pointer material since it is covered
    in chapter 5
Changed the structure and ordering of the chapter to teach all
    concepts required for dynamic arrays (array/pointer mingling,
    array memory layout, malloc) before seeing dynamic arrays.
    I was concerned with jumping straight into malloc as a solution
    before covering the "why" and "how".
Made regular edits (clarifications, wording)
Added a section on initializing arrays
Changed "2dNums" variable name to one that compiles
Added a clarification about row/column major layout not being fixed
Covered an array's memory layout more explicitly
Moved and reworked the pointer arithmetic section and tried to relate
    it more towards arrays in general than malloc
Introduced malloc as just an allocation mechanism, before tying it
    in with dynamic arrays to prevent it from looking like a magic bullet
Indtroduced sizeof more explicitly
Mentioned the second way of dynamically allocating 2d arrays

Changes to chapter 5:
Changed the title since "Passing Types" was a bit confusing
Minor formatting changes
Added an example of a pointer-to-pointer since I removed it from the
    arrays chapter

Punctuation fix in Operators.

### CODE REVIEW

**Comment by jrgold** [comment]
This is a pretty major change, and I'd like to get feedback on it. It's a pretty big restructure with some large removals and additions, so I'm happy to change anything.

The "Close pull request" button should not be write next to the "Comment" button where cancel would normally go.

File: arraysandpointers.tex [hunk 1]
```diff
@@ -1,144 +1,164 @@
...
+Remember that an array is really just a block of memory --- there is no better way to refer to it than by the memory address it starts at, and this is what C does.
+When we declare an array \lstinline!T array[n]!, the variable \lstinline!array! can be used as a pointer to the start of the array's memory, which is also the address of its first element.
+This means that where we were previously using \lstinline!&array[0]! we can now just use \lstinline!array!.
 
 \begin{codeblock}
-int* p = doSomething(); // Returns an int *
-int* p1 = p; // p1 points to the same thing p does.
+/* These are both the same */
+*(array + 1) = 1;
+array[1] = 1;
```
Comment (mon) [inline arraysandpointers.tex:102]: Perhaps mention that square brackets are the more common and preferred method for this course?

File: arraysandpointers.tex [hunk 2]
```diff
@@ -210,63 +191,67 @@ \subsection{Multidimensional Dynamic Array}
...
 
-Now you should know everything you need to allocate an array dynamically.
-Also note that you can always allocate an array with size 1 if you only need 1 element!
-In that case it is usually more common to use the \texttt{*} operator for dereferencing rather than \texttt{[]}.
+Since we are allocating multiple separate pieces of memory here, it is not enough to simply free \lstinline!arr!.
+We must be careful to free every element of \lstinline!arr! (the rows) before we free \lstinline!arr!.
+
+As with the nested arrays we saw earlier in the chapter, there is nothing special about our choice of \lstinline!arr! as an array of rows rather than an array of columns.
+
+The other commonly used method for dynamically-sized multidimensional arrays is to allocate a single block of memory large enough to hold every element, and calculate the indexes required for specific elements.
```
Comment (mon) [inline arraysandpointers.tex:201]: Perhaps an explanation on why you'd use this over the conceptually easier method

File: functions.tex [hunk 3]
```diff
@@ -101,7 +103,8 @@ \chapter{Functions and Passing types}
 int* i, j; // Warning: this does not mean what you think
 int *k; // the * can also go against the variable name
 int*p; // also legal. C doesn't care about spaces here
-int * q; // Now you're just being silly. (Still legal though).
+int * q; // Now you're just being silly. (Still legal though)
+int** m; // Pointers to pointers are okay too
 \end{codeblock}
 The last two lines are examples of legal code which is unnecessarily hard to read.
```
Comment (mon) [inline functions.tex]: No longer last 2 since you added the *\* example, probably just move m above the badly spaced ones

File: functions.tex [hunk 4]
```diff
@@ -117,6 +120,9 @@ \chapter{Functions and Passing types}
 the rest of the code you are writing.
 Most importantly, you need to conform with whatever style rules are in force for your organisation.
 
+Line~7 is an example of a pointer to a pointer.
+Since a pointers is a regular value, it must reside in memory, and so can be referred to with another pointer.
```
Comment (mon) [inline functions.tex]: s/pointers/pointer

**Comment by jrgold** [comment]
Fixed those up. For the multidimensional array styles, I only mentioned tradeoffs in creation/destruction/access, as small memory usage gains and performance gains seem a bit advanced at this stage.

File: arraysandpointers.tex [hunk 5]
```diff
@@ -1,144 +1,166 @@
...
+\section{Initializing an array}
+
+When you declare a variable (including arrays) in a function, C allocates memory for it but does not initialize the memory to any particular value.
+This means that if we just declare an array like we did above, it will contain gibberish elements.
+We could loop through the array and assign each element to something sensible, or we could use C's array initialization syntax:
+
+\begin{codeblock}
+int a[4] = { 0, 1, 5, 6 }; // All elements are initialized
+int b[4] = { 4, 3 }; // Remaining elements are initialized to 0
+int c[4] = { }; // All elements are initialized to 0
```
Comment (mitchellkrome) [inline arraysandpointers.tex:37]: test.c:29:16: warning: ISO C forbids empty initializer braces [-Wpedantic]
     int c[4] = {};

You always have to initialise 1 element. Usually you would do something like

int c[4] = { 0 };

**Comment by jrgold** [comment]
Whoops, forgot about that! I'll fix it up it a few hours when I have access to my computer.

**Comment by joelfenwick** [comment]
I'll fix it when I merge.

I may release a preliminary version today, we'll see what state it's in.

**Comment by jrgold** [comment]
Okay, sounds good.

**Pull request merged**